drivers/charger/charger_pca9422.c

```diff
@@ -0,0 +1,672 @@
+};
+
+static const struct linear_range i_fast_chg_ua_range[] = {
+	LINEAR_RANGE_INIT(2500, 2500, 0x0U, 0x7FU),
+};
+
+static int pca9422_charger_get_status(const struct device *dev, enum charger_status *status)
+{
+	const struct charger_pca9422_config *const config = dev->config;
+	uint8_t val;
+	int ret;
+
+	ret = mfd_pca9422_reg_read_byte(config->mfd, PCA9422_REG_CHARGER_0_STS, &val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (val & PCA9422_BIT_CHG_QUAL_OK) {
+		if (val & (PCA9422_BIT_PRECHARGE | PCA9422_BIT_FAST_CHARGE | PCA9422_BIT_CV_MODE |
+			   PCA9422_BIT_TOP_OFF)) {
+			*status = CHARGER_STATUS_CHARGING;
+		} else {
+			*status = CHARGER_STATUS_NOT_CHARGING;
+		}
+
+		/* Check charging done status */
+		ret = mfd_pca9422_reg_read_byte(config->mfd, PCA9422_REG_CHARGER_2_STS, &val);
+		if (ret < 0) {
+			return ret;
+		}
+		if (val & PCA9422_BIT_CHARGE_DONE) {
+			*status = CHARGER_STATUS_FULL;
+		}
+	} else {
+		*status = CHARGER_STATUS_NOT_CHARGING;
+	}
+
+	return 0;
+}
+
+static int pca9422_charger_get_online(const struct device *dev, enum charger_online *online)
+{
+	const struct charger_pca9422_config *const config = dev->config;
+	uint8_t val;
+	int ret;
+
+	ret = mfd_pca9422_reg_read_byte(config->mfd, PCA9422_REG_DEVICE_0_STS, &val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (val & PCA9422_BIT_VIN_OK) {
+		*online = CHARGER_ONLINE_FIXED;
+	} else {
+		*online = CHARGER_ONLINE_OFFLINE;
+	}
+
+	return 0;
+}
+
+static int pca9422_charger_set_constant_charge_current(const struct device *dev,
+						       uint32_t current_ua)
+{
+	const struct charger_pca9422_config *const config = dev->config;
+	struct charger_pca9422_data *data = dev->data;
+	uint16_t idx;
+	uint8_t val;
+	uint8_t num_ranges;
+	int ret;
+
+	current_ua = CLAMP(current_ua, CURRENT_STEP_2P5MA_MIN_UA, CURRENT_STEP_5P0MA_MAX_UA);
+
+	if ((data->chg_current_step == CHG_CURRENT_STEP_2P5MA) &&
+	    (current_ua > CURRENT_STEP_2P5MA_MAX_UA)) {
+		val = CHG_CURRENT_STEP_5P0MA;
+		val = FIELD_PREP(PCA9422_BIT_CHG_CURRENT_STEP, val);
+		ret = mfd_pca9422_reg_update_byte(config->mfd, PCA9422_REG_CHARGER_CNTL_1,
+						  PCA9422_BIT_CHG_CURRENT_STEP, val);
+		if (ret < 0) {
+			return ret;
+		}
+		data->chg_current_step = CHG_CURRENT_STEP_5P0MA;
+	}
+
+	num_ranges = ARRAY_SIZE(i_fast_chg_ua_range);
+	ret = linear_range_group_get_win_index(i_fast_chg_ua_range, num_ranges, current_ua,
+					       current_ua, &idx);
+	if (ret == -EINVAL) {
+		return ret;
+	}
+	ret = mfd_pca9422_reg_write_byte(config->mfd, PCA9422_REG_CHARGER_CNTL_3, (uint8_t)idx);
+	if (ret == 0) {
+		data->i_fast_chg_ua = current_ua;
+	}
+	return ret;
+}
+
+static int pca9422_charger_set_constant_charge_voltage(const struct device *dev,
+						       uint32_t voltage_uv)
+{
+	const struct charger_pca9422_config *const config = dev->config;
+	struct charger_pca9422_data *data = dev->data;
+	uint16_t idx;
+	uint8_t num_ranges;
+	int ret;
+
+	voltage_uv = CLAMP(voltage_uv, VBAT_REG_MIN_UV, VBAT_REG_MAX_UV);
```

Comment:
Should out of range values trigger EINVAL or ENOTSUP to alert the app developer? I'm not a maintainer, but had this thought. Some of the subsys drivers return errors, some do clamp to the range, I do not know the idiomatic way